Managed Media Source should only be available if airplay source alternative is present

jyavenard opened the pull request:
#### 0a9b7c59e3fa9131c40823fa108c5c9818269b08
<pre>
Managed Media Source should only be available if airplay source alternative is present
<a href="https://bugs.webkit.org/show_bug.cgi?id=256570">https://bugs.webkit.org/show_bug.cgi?id=256570</a>
rdar://109130836

Reviewed by Jer Noble.

When we add a Managed Media Source to a media element, it should per
MSE spec changed the MSE&apos;s readyState to &quot;open&quot; and fire the `sourceopen` event.
We want to make sure that the developer hasn&apos;t left anything unchecked,
it is expected that either an alternative source is set and compatible with
remote playback, or by explicitly disabling remote playback
(either by using an attribute or using Remote Playback API).

To achieve this, we do not move the MSE&apos;s readyState to &quot;open&quot; until
all those conditions have been achieved. And the readyState will remain
at `closed` until then. Should one of the condition become fullfilled
after the Managed Media Source has been created, it will be unblocked and
can be used as usual.

The behaviour is placed behind a preference.

Fly-by fix: There were three different ways to disable AirPlay, each
working independently from the other. We combine them all together that
should one method be used to disable AirPlay, it will be disabled regardless
of what the other methods are doing.

* LayoutTests/TestExpectations:
* LayoutTests/media/media-source/media-managedmse-airplay-expected.txt: Added.
* LayoutTests/media/media-source/media-managedmse-airplay.html: Added.
* LayoutTests/media/media-source/media-managedmse-bufferedchange.html: Amended to explicitly disable AirPlay
* LayoutTests/media/media-source/media-managedmse-memorypressure-inactive.html: Amended to explicitly disable AirPlay
* LayoutTests/media/media-source/media-managedmse-memorypressure.html: Amended to explicitly disable AirPlay
* LayoutTests/media/media-source/media-managedmse-streaming-atend.html: Amended to explicitly disable AirPlay
* LayoutTests/media/media-source/media-managedmse-streaming.html: Amended to explicitly disable AirPlay
* LayoutTests/media/media-source/media-managedmse.html: Amended to explicitly disable AirPlay
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
(WebCore::ManagedMediaSource::isOpen const):
* Source/WebCore/Modules/mediasource/ManagedMediaSource.h:
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::setPrivateAndOpen):
(WebCore::MediaSource::isOpen const):
(WebCore::MediaSource::isClosed const):
(WebCore::MediaSource::isEnded const):
(WebCore::MediaSource::openIfDeferredOpen):
(WebCore::MediaSource::readyState const):
(WebCore::MediaSource::onReadyStateChange):
* Source/WebCore/Modules/mediasource/MediaSource.h:
(WebCore::MediaSource::readyState const): Deleted.
* Source/WebCore/Modules/remoteplayback/RemotePlayback.cpp:
(WebCore::RemotePlayback::watchAvailability): Use utility method on mediaElement
(WebCore::RemotePlayback::cancelWatchAvailability): ditto
(WebCore::RemotePlayback::prompt): ditto
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::attributeChanged): If remote playback gets
disabled, move the Managed Media Source state to open if needed
(WebCore::isAllowedToLoadMediaURL): make it take a const element
(WebCore::HTMLMediaElement::isSafeToLoadURL const): make method const and make logging conditional
(WebCore::HTMLMediaElement::sourceWasAdded): if a playable alternative source is added
move the Managed Media Source state to open if needed
(WebCore::HTMLMediaElement::hasWirelessPlaybackTargetAlternative const):
(WebCore::HTMLMediaElement::isWirelessPlaybackTargetDisabled const):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/html/HTMLVideoElement.cpp:
(WebCore::HTMLVideoElement::attributeChanged):
* Source/WebKit/Shared/WebPreferencesDefaultValues.cpp:
(WebKit::defaultManagedMediaSourceNeedsAirPlay):
* Source/WebKit/Shared/WebPreferencesDefaultValues.h:

Canonical link: <a href="https://commits.webkit.org/264270@main">https://commits.webkit.org/264270@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/43d9f95d58901ae90ac8ae772d7afaf9a9e84bae

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6959 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7194 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7377 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8566 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7196 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6965 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/8458 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7132 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10105 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7082 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/7763 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6412 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8668 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/5136 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6325 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14103 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/5875 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/6812 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6416 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/9240 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/6535 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6890 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5649 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/7094 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6263 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/6235 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1660 "Passed tests") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1704 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10450 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/7294 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6645 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1800 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on LayoutTests/media/media-source/media-managedmse-airplay.html:
```diff
@@ -0,0 +1,78 @@
+            });
+
+            await sleepFor(100);
+            run('video.setAttribute("x-webkit-airplay", "deny")');
```
I would use the standard [disableRemotePlayback](https://w3c.github.io/remote-playback/#the-disableremoteplayback-attribute) attribute instead of `x-webkit-airplay` in the tests and in your commit message.

jyavenard replied:
I've added the ability to do either, and will use the Remote Playback API in the test.

eric-carlson commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -6493,6 +6504,10 @@ void HTMLMediaElement::playbackTargetPickerWasDismissed()
+#if ENABLE(MEDIA_SOURCE)
+    if (m_mediaSource)
+        m_mediaSource->openIfDeferredOpen();
+#endif
```
This method is called when AirPlay _listeners_ are added or removed. Why would that change the state of Managed Media Source?

Would it make sense to trigger a check to se if the source should be opened from `MediaSession::setWirelessVideoPlaybackDisabled`?

jyavenard replied:
this method is called whenever the Remote Playback API is disabled so I thought it would an easy path.
However, the JS binding simply set an attribute on the media element, to it's already covered by the `HTMLMediaElement::attributeChanged`

eric-carlson commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -6502,6 +6517,26 @@ void HTMLMediaElement::remoteHasAvailabilityCallbacksChanged()
+    return false;
+}
+
+bool HTMLMediaElement::hasWirelessPlaybackTargetDisabled() const
```
I think `isWirelessPlaybackTargetDisabled` would be more idiomatic.

jyavenard replied:
ok

eric-carlson commented on Source/WebCore/html/HTMLVideoElement.cpp:
```diff
@@ -151,12 +151,8 @@ void HTMLVideoElement::attributeChanged(const QualifiedName& name, const AtomStr
-            mediaSession().setWirelessVideoPlaybackDisabled(disabled);
-        }
+        if (name == webkitairplayAttr)
+            mediaSession().setWirelessVideoPlaybackDisabled(hasWirelessPlaybackTargetDisabled());
```
This shouldn't be changed, `x-webkit-airplay` is not a boolean attribute. 

Odd though it may be, `webkit-airplay` enables AirPlay and `x-webkit-airplay = deny` disables it.

jyavenard replied:
`isWirelessPlaybackTargetDisabled()` isn't a boolean checks
checks `equalLettersIgnoringASCIICase(attributeWithoutSynchronization(HTMLNames::webkitairplayAttr), "deny"_s)`

this change only moved the logic so it's all done in the same place.

However, it made little sense to enable AirPlay there, when the Remote Playback API would have disabled it. So I've combined all those together so that it's no consistent

eric-carlson commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -5259,6 +5265,11 @@ void HTMLMediaElement::sourceWasAdded(HTMLSourceElement& source)
     }
 
+#if ENABLE(MEDIA_SOURCE)
+    if (m_mediaSource)
+        m_mediaSource->openIfDeferredOpen();
```
Do we want to change the state if _any_ `<source>` is added, or only if it is one that is AirPlay-compatible?

jyavenard replied:
the end result is the same, and it allows to have management to when to allow MMS in a single place only

eric-carlson replied:
`openIfDeferredOpen` unconditionally sets the `readyState` to open, so won't the following enable MMS even though the new source element is useless?

    video.appendChild(document.createElement('source'))

jyavenard replied:
`openIfDeferredOpen` doesn't sets the readyState to open. It first checks if ManagedMediaSource::isOpen() returns true; which it won't unless the right conditions are matched.

jyavenard replied:
> video.appendChild(document.createElement('source'))

The code checks that there's a source, and that the src of that source is usable.

eric-carlson commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -6502,6 +6517,26 @@ void HTMLMediaElement::remoteHasAvailabilityCallbacksChanged()
+        String mimeType = source.attributeWithoutSynchronization(typeAttr);
+        ContentType contentType(mimeType);
+
+        if (contentType.containerType() == "application/vnd.apple.mpegurl"_s || contentType.containerType() == "application/x-mpegURL"_s)
```
AppleTV supports other containers as well, why only HLS?

jyavenard replied:
there's no MSE/MMS on ATV. But the rule is to enable MSE/MMS if we have an HLS alternative so that AirPlay works with MSE

jyavenard replied:
> AppleTV supports other containers as well, why only HLS?

Oh I see what you mean; I guess I could do any alternative source not containing a MSE URL.

jyavenard replied:
done

eric-carlson commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -827,6 +827,12 @@ void HTMLMediaElement::attributeChanged(const QualifiedName& name, const AtomStr
+        if (m_mediaSource && hasWirelessPlaybackTargetDisabled())
+            m_mediaSource->openIfDeferredOpen();
+        break;
+#endif
```
See the comment below about this strange attribute.

jyavenard replied:
I think the behaviour is kept following this change.
`isWirelessPlaybackTargetDisabled` checks that the value is set to "deny"

eric-carlson commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -2343,18 +2350,20 @@ static inline bool isAllowedToLoadMediaURL(HTMLMediaElement& element, const URL&
 {
     if (!url.isValid()) {
-        ERROR_LOG(LOGIDENTIFIER, url, " is invalid");
+        if (shouldLog)
+            ERROR_LOG(LOGIDENTIFIER, url, " is invalid");
```
Nit: this could be simplified to `ERROR_LOG_IF(shouldLog, LOGIDENTIFIER, url, " is invalid")`

eric-carlson commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -2343,18 +2350,20 @@ static inline bool isAllowedToLoadMediaURL(HTMLMediaElement& element, const URL&
         if (actionIfInvalid == Complain) {
             FrameLoader::reportLocalLoadFailed(frame.get(), url.stringCenterEllipsizedToLength());
-            ERROR_LOG(LOGIDENTIFIER, url , " was rejected by SecurityOrigin");
+            if (shouldLog)
+                ERROR_LOG(LOGIDENTIFIER, url , " was rejected by SecurityOrigin");
```
Ditto

eric-carlson commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -2363,13 +2372,15 @@ bool HTMLMediaElement::isSafeToLoadURL(const URL& url, InvalidURLAction actionIf
             if (frame)
                 FrameLoader::reportBlockedLoadFailed(*frame, url);
-            ERROR_LOG(LOGIDENTIFIER, url , " was rejected because the port is not allowed");
+            if (shouldLog)
+                ERROR_LOG(LOGIDENTIFIER, url , " was rejected because the port is not allowed");
```
Ditto

eric-carlson commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -2363,13 +2372,15 @@ bool HTMLMediaElement::isSafeToLoadURL(const URL& url, InvalidURLAction actionIf
 
     if (!isAllowedToLoadMediaURL(*this, url, isInUserAgentShadowTree())) {
-        ERROR_LOG(LOGIDENTIFIER, url, " was rejected by Content Security Policy");
+        if (shouldLog)
+            ERROR_LOG(LOGIDENTIFIER, url, " was rejected by Content Security Policy");
```
Ditto

jyavenard replied:
I missed those comments and had started the merge already. Will do in a follow up fix

jyavenard committed:
Managed Media Source should only be available if airplay source alternative is present
https://bugs.webkit.org/show_bug.cgi?id=256570
rdar://109130836

Reviewed by Jer Noble.

When we add a Managed Media Source to a media element, it should per
MSE spec changed the MSE's readyState to "open" and fire the `sourceopen` event.
We want to make sure that the developer hasn't left anything unchecked,
it is expected that either an alternative source is set and compatible with
remote playback, or by explicitly disabling remote playback
(either by using an attribute or using Remote Playback API).

To achieve this, we do not move the MSE's readyState to "open" until
all those conditions have been achieved. And the readyState will remain
at `closed` until then. Should one of the condition become fullfilled
after the Managed Media Source has been created, it will be unblocked and
can be used as usual.

The behaviour is placed behind a preference.

Fly-by fix: There were three different ways to disable AirPlay, each
working independently from the other. We combine them all together that
should one method be used to disable AirPlay, it will be disabled regardless
of what the other methods are doing.

* LayoutTests/TestExpectations:
* LayoutTests/media/media-source/media-managedmse-airplay-expected.txt: Added.
* LayoutTests/media/media-source/media-managedmse-airplay.html: Added.
* LayoutTests/media/media-source/media-managedmse-bufferedchange.html: Amended to explicitly disable AirPlay
* LayoutTests/media/media-source/media-managedmse-memorypressure-inactive.html: Amended to explicitly disable AirPlay
* LayoutTests/media/media-source/media-managedmse-memorypressure.html: Amended to explicitly disable AirPlay
* LayoutTests/media/media-source/media-managedmse-streaming-atend.html: Amended to explicitly disable AirPlay
* LayoutTests/media/media-source/media-managedmse-streaming.html: Amended to explicitly disable AirPlay
* LayoutTests/media/media-source/media-managedmse.html: Amended to explicitly disable AirPlay
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
(WebCore::ManagedMediaSource::isOpen const):
* Source/WebCore/Modules/mediasource/ManagedMediaSource.h:
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::setPrivateAndOpen):
(WebCore::MediaSource::isOpen const):
(WebCore::MediaSource::isClosed const):
(WebCore::MediaSource::isEnded const):
(WebCore::MediaSource::openIfDeferredOpen):
(WebCore::MediaSource::readyState const):
(WebCore::MediaSource::onReadyStateChange):
* Source/WebCore/Modules/mediasource/MediaSource.h:
(WebCore::MediaSource::readyState const): Deleted.
* Source/WebCore/Modules/remoteplayback/RemotePlayback.cpp:
(WebCore::RemotePlayback::watchAvailability): Use utility method on mediaElement
(WebCore::RemotePlayback::cancelWatchAvailability): ditto
(WebCore::RemotePlayback::prompt): ditto
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::attributeChanged): If remote playback gets
disabled, move the Managed Media Source state to open if needed
(WebCore::isAllowedToLoadMediaURL): make it take a const element
(WebCore::HTMLMediaElement::isSafeToLoadURL const): make method const and make logging conditional
(WebCore::HTMLMediaElement::sourceWasAdded): if a playable alternative source is added
move the Managed Media Source state to open if needed
(WebCore::HTMLMediaElement::hasWirelessPlaybackTargetAlternative const):
(WebCore::HTMLMediaElement::isWirelessPlaybackTargetDisabled const):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/html/HTMLVideoElement.cpp:
(WebCore::HTMLVideoElement::attributeChanged):
* Source/WebKit/Shared/WebPreferencesDefaultValues.cpp:
(WebKit::defaultManagedMediaSourceNeedsAirPlay):
* Source/WebKit/Shared/WebPreferencesDefaultValues.h:

Canonical link: https://commits.webkit.org/264270@main
